### PR TITLE
find_and_replace network filter

### LINKED ISSUE #4462: Generic string (and bytes) search and replace network filter
*Description*:
I'm proposing to create a generic string (and bytes) search and replace network filter, that can replace strings and bytes in the payload at the beginning of a downstream connection (readFilter) as well as when sending the response (writeFilter). Sometimes one needs to massage incoming and outgoing payloads just a little bit (e.g. legacy applications or missing control over client behavior).

Since matching arbitrary strings across different packets/buffers is hard, I'd limit this to a certain amount of data at the beginning of the payload, when the read/write filter chain is invoked. The filter would wait until a certain amount of data has been accumulated and then would try to search and replace. This amount of data could be specified as a string/bytes sequence (e.g. "\r\n\r\n" to catch all HTTP/1 headers) or as an amount of bytes. Should the payload never include the sequence or reach the amount of bytes, then we could continue down the filter chain after some configurable timeout.

Search and replace could be done via statically matched strings (or bytes in hex?) or via regular expressions.

Example:

```
filters:
  - name: search_replace
    config:
      read:
        search: "old"
        replace: "new"
        mode: string
        wait_for_bytes: 50
        timeout_in_ms: 2000
      write:
        search: "(.*)new(.*)"
        replace: "$1old$2"
        mode: regex
        wait_for_string: "\n"
        timeout_in_ms: 1000
```
(another `mode` could be `bytes`, with `search` and `replace` being hex values)

With this configuration a client could send the following payload:
```
blah blah super old proprietary protocol request blah blah
```
And the filter would modify this to
```
blah blah super new proprietary protocol request blah blah
```
before passing it on.

On the way back downstream: 
```
blah blah dynamic 123 super new proprietary protocol response unique value blah blah\n
```
would be turned into
```
blah blah dynamic 123 super old proprietary protocol response unique 567 value blah blah\n
```

(You get the picture...)

What do we think? Worthwhile to have this functionality? I'd volunteer to work on this.

### PR DESCRIPTION
**Pull request by bmetzdorf**
Signed-off-by: Bjoern Metzdorf <bmetzdorf@apple.com>

*Description*:
I'm proposing to create a generic string (and bytes) find and replace network filter, that can replace strings and bytes in the payload at the beginning of a downstream connection (readFilter) as well as when sending the response (writeFilter). Sometimes one needs to massage incoming and outgoing payloads just a little bit (e.g. legacy applications or missing control over client behavior).

For more details please see https://github.com/envoyproxy/envoy/issues/4462

This is a very simple first version which lacks proper and proto'ized configuration and is meant to gather feedback.

*Risk Level*: low

*Testing*: ran all tests locally and they're all green :)

*Docs Changes*: TBD

*Release Notes*: TBD

Fixes #4462 

@mattklein123 @ggreenway

### CODE REVIEW

**Comment by rshriram** [comment]
does this apply to TLS data received from downstream? can I use this to rewrite SNI?

**Comment by bmetzdorf** [comment]
@rshriram No, this is past TLS. From what I can see you're looking for a listener filter here, similar to the [TLS Inspector](https://www.envoyproxy.io/docs/envoy/latest/configuration/listener_filters/tls_inspector).

**Comment by rshriram** [comment]
Let me rephrase. In an situation where I am doing SNI passthrough (not terminating TLS), envoy sees it simply as a byte stream on a TCP connection. So it should be possible to rewrite the client hello before forwarding to the upstream cluster, for as far as Envoy is concerned, its just a bunch of bytes on tcp proxy.

**Comment by junr03** [comment]
@ggreenway can you take a pass to provide early feedback on the direction?

**Comment by bmetzdorf** [comment]
@rshriram Ok, I see what you mean. But AFAIR all TLS handshake protocol messages are tamper-proof (using the `Finished` message in both TLS 1.2 and 1.3) and any middlebox modification will immediately lead to an invalid handshake.

**Comment by rshriram** [comment]
SNI is sent in clear text. It can be rewritten.

**Comment by bmetzdorf** [comment]
@rshriram Although it is sent in cleartext it is still protected. If you do rewrite it then the TLS handshake will fail because both TLS endpoints maintain hashes of all handshake protocol messages, which have to be identical for a successful handshake.

Does that make sense? Can you maybe explain your use case?

**Comment by bmetzdorf** [comment]
@ggreenway Should I be worried about the tsan CI failure or just trigger CI again?

**Comment by bmetzdorf** [comment]
Ok, the tsan build is also green now.

**Comment by bmetzdorf** [comment]
@rshriram You're welcome! Still care to explain your use case? Maybe we can find another solution for it?

File: source/extensions/filters/network/find_and_replace/config.cc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    ConfigConstSharedPtr filter_config(
+        std::make_shared<Config>(dynamic_cast<const Envoy::ProtobufWkt::Struct&>(proto_config)));
+
+    return [filter_config](Network::FilterManager& filter_manager) -> void {
+      filter_manager.addFilter(std::make_shared<Filter>(filter_config));
+    };
+  }
+
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return std::make_unique<Envoy::ProtobufWkt::Struct>();
```
Comment (ggreenway) [inline source/extensions/filters/network/find_and_replace/config.cc]: Let's start working out what the proto config for this will be.

**Comment by bmetzdorf** [comment]
So as a first step I proto'ized the existing config. This config proposal here is similar to the one in #4462 but it does make `mode` a more top-level thing: 

```
filters:
  - name: envoy.filters.network.find_and_replace
    config:
      read:
        mode: string
          search: "old"
          replace: "new"
          wait_for_string: "\r\n\r\n"
          timeout_in_ms: 2000
      write:
        mode: bytes
          search: "CC00FFFFEE"
          replace: "11EEAA"
          wait_for_bytes: 50
          timeout_in_ms: 1000
```
`read` and `write` will probably not change anytime soon. And `mode` can be a new `one-of`. Every `mode` then manages its own configuration and thus we can stay flexible and introduce new modes whenever we want to add or change features.

I think that given the constraints around regex performance and general unbounded string search across multiple packets, where the search string length is not known at search time, it might make sense to for now not support the regex mode.

**Comment by mattklein123** [comment]
I will help review this week, since I think there is going to be some matching overlap with tap/dump and I want to think about how we can share matchers in common code as much as possible.

File: api/envoy/config/filter/network/find_and_replace/v2alpha1/find_and_replace.proto [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import "validate/validate.proto";
+import "gogoproto/gogo.proto";
+
+// [#protodoc-title: Find and Replace]
+// Find and Replace Network Filter :ref:`configuration overview
+// <config_network_filters_find_and_replace>`.
+
+// Find and Replace network filter config.
+//
+message FindAndReplace {
```
Comment (mattklein123) [inline api/envoy/config/filter/network/find_and_replace/v2alpha1/find_and_replace.proto]: Can you add more complete docs here so that it's easier to understand at the config layer how this filter would be used? (Obviously in the final PR we will need complete docs, but let's start here w/ the protos). I suspect there is going to be overlap with tap/dump matching, and I want to make sure we future proof the matching part of this so that we can use it in both places.

Comment (bmetzdorf) [reply api/envoy/config/filter/network/find_and_replace/v2alpha1/find_and_replace.proto]: Done (at least for this initial config).

File: api/envoy/config/filter/network/find_and_replace/v2alpha1/find_and_replace.proto [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  // Optional string to find in data from the downstream side (read filter).
+  // Finding the string is currently only supported from the very beginning of the data on a given
+  // connection, meaning the string has to match the beginning of the data. If empty (or not
+  // specified), then :ref:`input_rewrite_to
+  // <envoy_api_field_config.filter.network.find_and_replace.v2alpha1.FindAndReplace.input_rewrite_to>`
+  // effectively becomes an insertion instead of a replacement string. Defaults to empty.
+  string input_rewrite_from = 1;
```
Comment (mattklein123) [inline api/envoy/config/filter/network/find_and_replace/v2alpha1/find_and_replace.proto]: String matcher here? Same on the match below?

Also, to better future proof this, can we think about logically breaking this up into multiple different messages? For example:
1) Input matcher
2) Input translation
3) Output matcher
4) Output translation

I think we can then have common matching and translation constructs for both input and output, possible with some well placed oneofs to make it clear that we can add additional matching options in the future?

Comment (bmetzdorf) [reply api/envoy/config/filter/network/find_and_replace/v2alpha1/find_and_replace.proto]: I think I need some more guidance here. The issue I'm facing is that I don't know how to reasonably use the StringMatcher in conjunction with `Buffer::Instance::search()`. I guess I could utilize `Buffer::Instance::toString()` and call `Matchers::StringMatcher::match()` with that, but how would that efficiently work for any matching mode besides `prefix`?

For the `exact` match: To truly do an exact match I'd have to wait for all incoming data ever., which defeats the purpose here.

The `suffix` and `regex` matches have a similar issue: How do I know when to stop buffering and execute the comparison, since I don't know the length of the search string?

`prefix` does make sense, since I know the string length (which is the current functionality of this PR).

For example, let's assume incoming data from downstream is split across 2 packets. The first packet contains `abc` and the second contains `def`.
- `prefix` StringMatcher using `abcd` would work after receiving the second packet (likely the second `onData()` invocation).
- `exact` StringMatcher using `abcd` would never match, since both `onData()` calls never just contain `abcd`, and even if they did, then it would still be wrong, since it is an incoming data stream and I don't know how many more `onData()` invocations will happen.
- `suffix` and `regex` StringMatcher using `abcd` have nearly the same issue as `exact`: I don't know when the incoming data stream ends. (And currently it seems as if we're using `regex_match`, which works on the whole string, so it is facing the same issues as `exact`)

For any mode besides `prefix` we would have to know how much incoming data to wait for.

Does that make any sense or am I completely off here?

**Comment by bmetzdorf** [comment]
I’ll think about something. I’ll have some time after thanksgiving..
